### PR TITLE
Allow factors in Linear regression

### PR DESCRIPTION
**Pull request by vandenman**
Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/1155
Fixes https://github.com/jasp-stats/jasp-issues/issues/1022

This got a lot more involved than I hoped for, but the analysis assumed everywhere that the number of covariates is equal to the number of parameter estimates. Once you introduce factors, you get a parameter for the number of factor levels - 1, so this assumption is violated.

I fixed all bugs that I encountered, but I think that this analysis should be checked properly because in particular interaction effects with categorical predictors can lead to really error-prone models.

### CODE REVIEW

**Review by Kucharssim — COMMENTED** [review]
Just one thing in the code, I still need to test it though

File: R/regressionlinear.R [hunk 1]
```diff
@@ -874,7 +929,7 @@ RegressionLinear <- function(jaspResults, dataset = NULL, options) {
 .linregTryToAddOnePredictor <- function(prevModel = NULL, dependent, candidatePredictors, predictorsInNull, data, options, weights) {
   fValues <- numeric(length(candidatePredictors))
   pValues <- numeric(length(candidatePredictors))
-
+browser()
```
[the remark this answers is not in the document]
Comment (vandenman) [reply R/regressionlinear.R:932]: Ah shit I forgot about that one. I disabled stepwise methods for factors so that's my (weak) excuse.

**Comment by Kucharssim** [comment]
On the latest mac nightly (https://github.com/jasp-stats/jasp-desktop/commit/b7d2d00ea8694c3a034115ea62117d64216b7f57) I get this: 
Screenshot 2020-12-01 at 11 04 05

@vandenman @JorisGoosen Am I right that this is a problem of the nightly, not the module, right? I think `jaspBase` *should* export this function...

**Comment by vandenman** [comment]
@Kucharssim you can't test this on a nightly, as it requires the latest version of `jaspBase`. So you'll need to build jasp (and run `git submodule update --remote` to pull in the latest changes from jaspBase beforehand)

**Comment by JorisGoosen** [comment]
I'll update the modules to the latest version and build new nightlies.

**Comment by JohnnyDoorn** [comment]
while you're doing that, could you also update the flatpak beta? :D

**Comment by JorisGoosen** [comment]
Haha, yeah sure. But that'll take till tomorrow to be updated.

File: R/regressionlinear.R [hunk 2]
```diff
@@ -1419,115 +1485,145 @@ RegressionLinear <- function(jaspResults, dataset = NULL, options) {
...
+      lowerPredInt <- (intercept + slope * xVar) - qt(ciPred, dfRes) * sePred
+
+      predictionIntervalLines <- if (is.factor(xVar)) {
+        ggplot2::geom_errorbar(
+          data = data.frame(x = levels(xVar), ymax = upperPredInt, ymin = lowerPredInt),
+          mapping = ggplot2::aes(x = x, ymax = ymax, ymin = ymin),
+          colour = "darkgreen", linetype = "longdash"
+        )
+      } else {
+        ggplot2::geom_line(data = data.frame(x = xVar, y = c(upperConfInt, lowerConfInt), g = rep(1:2, c(length(upperConfInt), length(lowerConfInt)))),
```
Comment (Kucharssim) [inline R/regressionlinear.R:1608]:
```suggestion
        ggplot2::geom_line(data = data.frame(x = xVar, y = c(upperPredInt, lowerPredInt), g = rep(1:2, c(length(upperPredInt), length(lowerPredInt)))),
```

Comment (Kucharssim) [reply R/regressionlinear.R:1608]: Forgot to change confidence -> prediction intervals, so that now you draw the same interval twice.

Comment (Kucharssim) [reply R/regressionlinear.R:1608]: I wonder why this was not caught by the unit test though, I though @maltelueken added new tests for prediction and confidence intervals in this PR: https://github.com/jasp-stats/jasp-desktop/pull/4220...

Ah I see, the change to the unit test was not copied to this module though: https://github.com/jasp-stats/jaspRegression/commit/9ec4037fdffa42d26d4bf46b38239c831e52af29

Would be good to add it back I think. @vandenman do you wanna do it here or should we just tackle this later?

**Review by Kucharssim — CHANGES_REQUESTED** [review]
Looks good, even feels relatively robust when bootstrapping with weird factor predictors, I expected worse (given our experience in logistic regression).

There is one small bug I found, see the code suggestion.

File: R/regressionlinear.R (hunk 2 above)
Comment (vandenman) [reply R/regressionlinear.R:1608]: Good catch! I was already wondering why the two lines were the same.. I'll add the unit test back in this PR.

Comment (vandenman) [reply R/regressionlinear.R:1608]: actually, it looks like there are some tests for these plots already:

https://github.com/jasp-stats/jasp-desktop/blob/22c87484bb28c7eb3b5ace15ae269942a356d7d5/JASP-Tests/R/tests/testthat/test-regressionlinear.R#L444-L456

and also here:

https://github.com/jasp-stats/jaspRegression/blob/f85592d261a095cbc0c2279611c49288965779b8/tests/testthat/test-regressionlinear.R#L444-L456

in both stable and this module, the unit tests files are identical for linear regression... In my next commit I'll re-enabled some of the partial plots, but these are already commented out in stable.

**Review by Kucharssim — APPROVED** [review]
It's fine with me now

**Comment by vandenman** [comment]
@JohnnyDoorn any timeline on when you can give feedback?

**Review by JohnnyDoorn — APPROVED** [review]

**Pull request merged**